src/garage/envs/environment.py

```diff
@@ -0,0 +1,189 @@
+"""Base Garage Environment API."""
+
+import abc
+
+
+class Environment(abc.ABC):
+    """The main API for garage environments.
+
+    The main public API methods are:
+
+        step
+        reset
+        render
+        visualize
+        close
+
+    And set the following attributes:
+
+        action_space: The Space object corresponding to valid actions
+        observation_space: The Space object corresponding to valid observations
+        reward_range: A tuple corresponding to the min and max possible rewards
+
+    Make sure your environment is pickle-able:
+        Garage pickles the environment via the `pickle` or `cloudpickle` module
+        to save snapshots of the experiment. However, some environments may
+        contain attributes that are not pickle-able (e.g. a client-server
+        connection). In such cases, override `__getstate__()` and
+        `__setstate()__` to add your custom pickle logic.
+
+    Note: a default reward range set to [-inf,+inf] already exists. Set it if
+        you want a narrower range.
+
+    """
+
+    # Set this in SOME subclasses
+    metadata = {'render.modes': []}
```

Comment:
why have this mutable metadata field? will this ever contain keys other than `render.modes`? Why not instead just have an abstract property `render_modes`